thingsboard_gateway/tb_utility/tb_gateway_remote_configurator.py

```diff
@@ -12,288 +12,559 @@
-                    self.__new_configuration["thingsboard"]["thingsboard"]["remoteShell"] = True
-                    
-                self.__new_logs_configuration = b64decode(self.__new_general_configuration_file.pop("logs")).decode('UTF-8').replace('}}', '\n')
-                if self.__old_configuration != decoded_configuration:
-                    LOG.info("Remote configuration received: \n %s", decoded_configuration)
-                    result = self.__process_connectors_configuration()
-                    self.in_process = False
-                    if result:
-                        self.__old_configuration = self.__new_configuration
-                        return True
-                    else:
-                        return False
-                else:
-                    LOG.info("Remote configuration is the same.")
-            else:
-                LOG.error("Remote configuration is already in processing")
-                return False
-        except Exception as e:
-            self.in_process = False
-            LOG.exception(e)
+    @property
+    def general_configuration(self):
+        return self._config.get('thingsboard', {})
+
+    @general_configuration.setter
+    def general_configuration(self, config):
+        self._config['thingsboard'].update(config)
+
+    @property
+    def storage_configuration(self):
+        return self._config.get('storage', {})
+
+    @storage_configuration.setter
+    def storage_configuration(self, config):
+        self._config['storage'].update(config)
+
+    @property
+    def grpc_configuration(self):
+        return self._config.get('grpc', {})
+
+    @grpc_configuration.setter
+    def grpc_configuration(self, config):
+        self._config.get('grpc', {}).update(config)
+
+    @property
+    def connectors_configuration(self):
+        connectors = self._config.get('connectors', [])
+        for connector in connectors:
+            connector.pop('config_updated', None)
+            connector.pop('config_file_path', None)
+        return connectors
+
+    def _get_active_connectors(self):
+        return [connector['name'] for connector in self.connectors_configuration]
+
+    def _get_tb_gateway_general_config_for_save(self):
+        """
+        Method used only for saving data to conf files
+        !!!Don't use it for sending data to TB (use `_get_tb_gateway_general_config_for_remote` instead)!!!
+        """
+
+        connectors_config = [
+            {'type': connector['type'], 'name': connector['name'], 'configuration': connector['configuration']} for
+            connector in self.connectors_configuration]
+
+        return {
+            'thingsboard': self.general_configuration,
+            'storage': self.storage_configuration,
+            'grpc': self.grpc_configuration,
+            'connectors': connectors_config
+        }
+
+    def _get_tb_gateway_general_config_for_remote(self):
+        """
+        Method used only for saving data to TB
```

Comment:
This method returns general configuration in format that should be used only for sending configuration to the server.